src/graphics/program-lib/chunks/ltc.frag

```diff
@@ -0,0 +1,447 @@
+
+	dSphereRadius = radius;
+
+	vec3 f = normalize(lightPos-view_position);
+	vec3 w = normalize(cross(f, halfHeight));
+	vec3 h = normalize(cross(f, w));
+
+	coords.coord0 = lightPos + w*radius - h*radius;
+	coords.coord1 = lightPos - w*radius - h*radius;
+	coords.coord2 = lightPos - w*radius + h*radius;
+	coords.coord3 = lightPos + w*radius + h*radius;
+	return coords;
+}
+
+// used for LTC LUT texture lookup
+vec2 dLTCUV;
+#ifdef CLEARCOAT
+vec2 ccLTCUV;
+#endif
+vec2 getLTCLightUV(float tGlossiness, vec3 tNormalW)
+{
+	float roughness = max((1.0 - tGlossiness) * (1.0 - tGlossiness), 0.001);
+	return LTC_Uv( tNormalW, dViewDirW, roughness );
+}
+
+//used for energy conservation and to modulate specular
+vec3 dLTCSpecFres;
+#ifdef CLEARCOAT
+vec3 ccLTCSpecFres;
+#endif
+vec3 getLTCLightSpecFres(vec2 uv, vec3 tSpecularity)
+{
+	vec4 t2 = texture2D( areaLightsLutTex2, uv );
+
+	#ifdef HAS_R8_G8_B8_A8_LUTS
+	t2 *= vec4(0.693103,1,1,1);
+	t2 += vec4(0.306897,0,0,0);
+	#endif
+
+	return tSpecularity * t2.x + ( vec3( 1.0 ) - tSpecularity) * t2.y;
+}
+
+void calcLTCLightValues()
+{
+	dLTCUV = getLTCLightUV(dGlossiness, dNormalW);
+	dLTCSpecFres = getLTCLightSpecFres(dLTCUV, dSpecularityNoFres); 
+
+#ifdef CLEARCOAT
+	ccLTCUV = getLTCLightUV(ccGlossiness, ccNormalW);
+	ccLTCSpecFres = getLTCLightSpecFres(ccLTCUV, ccSpecularityNoFres);
+#endif
+}
+
+void calcRectLightValues(vec3 lightPos, vec3 halfWidth, vec3 halfHeight)
+{
+	dLTCCoords = getLTCLightCoords(lightPos, halfWidth, halfHeight);
+	calcLTCLightValues();
+}
+void calcDiskLightValues(vec3 lightPos, vec3 halfWidth, vec3 halfHeight)
+{
+	calcRectLightValues(lightPos, halfWidth, halfHeight);
+}
+void calcSphereLightValues(vec3 lightPos, vec3 halfWidth, vec3 halfHeight)
+{
+	dLTCCoords = getSphereLightCoords(lightPos, halfWidth, halfHeight);
+	calcLTCLightValues();
+}
+
+// An extended version of the implementation from
+// "How to solve a cubic equation, revisited"
+// http://momentsingraphics.de/?p=105
+vec3 SolveCubic(vec4 Coefficient)
+{
+	float pi = 3.14159;
+    // Normalize the polynomial
+    Coefficient.xyz /= Coefficient.w;
+    // Divide middle coefficients by three
+    Coefficient.yz /= 3.0;
+
+    float A = Coefficient.w;
+    float B = Coefficient.z;
+    float C = Coefficient.y;
+    float D = Coefficient.x;
+
+    // Compute the Hessian and the discriminant
+    vec3 Delta = vec3(
+        -Coefficient.z*Coefficient.z + Coefficient.y,
+        -Coefficient.y*Coefficient.z + Coefficient.x,
+        dot(vec2(Coefficient.z, -Coefficient.y), Coefficient.xy)
+    );
+
+    float Discriminant = dot(vec2(4.0*Delta.x, -Delta.y), Delta.zy);
+
+    vec3 RootsA, RootsD;
+
+    vec2 xlc, xsc;
+
+    // Algorithm A
+    {
+        float A_a = 1.0;
+        float C_a = Delta.x;
+        float D_a = -2.0*B*Delta.x + Delta.y;
+
+        // Take the cubic root of a normalized complex number
+        float Theta = atan(sqrt(Discriminant), -D_a)/3.0;
+
+        float x_1a = 2.0*sqrt(-C_a)*cos(Theta);
```

Comment:
spacing

Reply:
I will add spaces - NB this code was copied directly from the original implementation (linked at the top of the file) - spacing between `+` and `-`  is different than `*` and `/` - resulting in IMO very readable code - e.g `A*B + C*D` vs our standard which is `A * B + C * D`

Reply:
I would say stick to the standard. If you want to argue for a change to the standard, then we migrate the whole codebase. We should avoid deviations made on a whim.